### PR TITLE
Fixes #34: Replace Processing.org's applet

### LINKED ISSUE #34: Substitute Processing implementation
Currently the app is build using Processing, which is not the best choice performance wise. Please propose and implement an alternative.

### PR DESCRIPTION
**Pull request by SKrPl**
Solves issue: https://github.com/fossasia/loklak_wok_android/issues/34

Uses loklak's "suggest" API to fetch queries to scrape, these queries are sent to twitter.js (transpiled and bundled) as a raw resource, for scraping. The scraped data is sent back to UI thread and displayed in a RecyclerView. The scraped tweets are pushed back to loklak server using "push" API.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -36,28 +36,23 @@ dependencies {
...
     compile 'com.squareup.retrofit2:converter-gson:2.3.0'
     compile 'com.squareup.retrofit2:adapter-rxjava2:2.3.0'
-
     compile 'com.jakewharton:butterknife:8.6.0'
     annotationProcessor 'com.jakewharton:butterknife-compiler:8.6.0'
-
     compile 'io.reactivex.rxjava2:rxjava:2.0.5'
     compile 'io.reactivex.rxjava2:rxandroid:2.0.1'
     compile 'com.jakewharton.rxbinding2:rxbinding:2.0.0'
+    compile 'com.github.LiquidPlayer:LiquidCore:0.2.2'
```
Comment (abishekvashok) [inline app/build.gradle]: Use implementation instead. `compile` is depreciated.

**Review by abishekvashok — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 2]
```diff
@@ -5,7 +5,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.3.2'
+        classpath 'com.android.tools.build:gradle:2.3.3'
```
Comment (abishekvashok) [inline build.gradle:8]: Pls use the latest one: 4.0

**Review by abishekvashok — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 2 above)
Comment (SKrPl) [reply build.gradle:8]: How do I update?

**Comment by SKrPl** [comment]
@Orbiter @sudheesh001 @mariobehling @daminisatya @abishekvashok @Achint08 @djmgit @hemantjadon @kavithaenair @SarishiNoHara @singhpratyush @vibhcool Please review it.

device-2017-06-30-060008

File: app/build.gradle [hunk 3]
```diff
@@ -19,7 +19,7 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
     }
-    // This is important, it will run lint checks but won't abort build
+    // This is compilet, it will run lint checks but won't abort build
```
Comment (kavithaenair) [inline app/build.gradle]: Typo.

File: build.gradle (hunk 2 above)
Comment (abishekvashok) [reply build.gradle:8]: Change 2.3.3 to 4.0
IDE will download the grade version for you.

File: app/build.gradle [hunk 4]
```diff
@@ -52,6 +53,7 @@ dependencies {
 
     compile 'com.squareup.retrofit2:retrofit:2.3.0'
     compile 'com.squareup.retrofit2:converter-gson:2.3.0'
+    compile 'com.squareup.okhttp3:logging-interceptor:3.7.0'
```
Comment (abishekvashok) [inline app/build.gradle:56]: Pls use implementation instead as compile is deprecated.

File: app/src/main/res/layout/fragment_tweet_harvesting.xml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        android:padding="@dimen/network_error_padding"
+        android:gravity="center"
+        android:layout_centerHorizontal="true"
+        android:layout_centerVertical="true"
+        android:text="@string/network_error_retry"
+        android:textSize="@dimen/message_font_size"
+        android:textColor="@android:color/black"
+        android:visibility="gone"/>
+
+</RelativeLayout>
```
Comment (abishekvashok) [inline app/src/main/res/layout/fragment_tweet_harvesting.xml]: Always add a newline at eof

File: app/src/main/res/menu/menu_tweet_harvesting.xml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <item
+        android:title="Search"
+        app:showAsAction="always"
+        android:id="@+id/search_tweets"
+        android:icon="@drawable/ic_search_24dp"/>
+
+</menu>
```
Comment (abishekvashok) [inline app/src/main/res/menu/menu_tweet_harvesting.xml]: New line

**Review by abishekvashok — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/row_harvested_tweet.xml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                android:layout_height="wrap_content"
+                android:id="@+id/harvested_tweet_text"
+                android:maxLines="2"
+                android:ellipsize="end"/>
+
+        </LinearLayout>
+
+    </LinearLayout>
+
+</android.support.v7.widget.CardView>
```
Comment (kavithaenair) [inline app/src/main/res/layout/row_harvested_tweet.xml]: New line.

File: app/src/main/res/drawable/ic_search_24dp.xml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+        android:width="24dp"
+        android:height="24dp"
+        android:viewportWidth="24.0"
+        android:viewportHeight="24.0">
+    <path
+        android:fillColor="@color/colorAccent"
+        android:pathData="M15.5,14h-0.79l-0.28,-0.27C15.41,12.59 16,11.11 16,9.5 16,5.91 13.09,3 9.5,3S3,5.91 3,9.5 5.91,16 9.5,16c1.61,0 3.09,-0.59 4.23,-1.57l0.27,0.28v0.79l5,4.99L20.49,19l-4.99,-5zM9.5,14C7.01,14 5,11.99 5,9.5S7.01,5 9.5,5 14,7.01 14,9.5 11.99,14 9.5,14z"/>
```
Comment (kavithaenair) [inline app/src/main/res/drawable/ic_search_24dp.xml:8]: Avoid lengthy lines.

**Review by singhpratyush — CHANGES_REQUESTED** [review]
The `twitter.js` file is huge. I think that webpack should be configured in a way that it produces a compressed version of bundled `js` (see [here](https://hackernoon.com/optimising-your-application-bundle-size-with-webpack-e85b00bab579) for help).

Also, to avoid the pain of updating everytime `loklak_js_scrapers` change, we should create an autobuild at separate branch using Travis and let apps use the resources from there.

@Achint08 @hemantjadon @SKrPl: What are your views on this?

**Comment by SKrPl** [comment]
@singhpratyush I have used `-p` flag for bundling (results in 390 lines), I will look into additionals provided in the blog.
Can you elaborate a bit on "we should create an autobuild at separate branch using Travis and let apps use the resources from there"?
The `twitter.js` file can't be kept on a separate branch, it needs to be on the same branch, so LiquidCore can parse it for execution.

File: build.gradle (hunk 2 above)
Comment (SKrPl) [reply build.gradle:8]: To use version `4.0` the users need to have Android Studio 3.0+ which is still a beta release, most of the users stick to a stable release - current stable release 2.3.3, and Android Studio below 3.x can't use the new gradle versions (links: [Android dev](https://developer.android.com/studio/preview/features/new-android-plugin-migration.html) and [SO](https://stackoverflow.com/questions/44580320/how-to-use-gradle-4-0-in-android-studio-2-3)). So, I think its better to stick with `2.3.3` till Android Studio 3.0 becomes a stable release :smile: .
@abishekvashok What's your view on this ?

File: app/build.gradle (hunk 4 above)
Comment (SKrPl) [reply app/build.gradle:56]: Can't be used with the current stable version of Android Studio.

**Comment by singhpratyush** [comment]
> [<span>@singhpratyush</span>](https://github.com/singhpratyush) I have used `-p` flag for bundling, I will look into additionals provided in the blog.

:+1: 

> Can you elaborate a bit on "we should create an autobuild at separate branch using Travis and let apps use the resources from there"?
The `twitter.js` file can't be kept on a separate branch, it needs to be on the same branch, so LiquidCore can parse it for execution.

I apologise for the abstract mention, please let me explain in detail about what flow I am proposing - 

1. The successful build at master of `loklak_scrapers_js` builds several bundled `js` and pushes it to another branch (say `bundles`) with proper versioning (maybe `<bundle_name>_v<travis_build_number>.<extension>`).
  We can also think about making automatic releases for this.

2. The app tries to see if it already has the bundle. If no, it can request the branch/release for latest version of the bundle and parse it with `LiquidCore`.
  Automatic, periodic updates would also be required.

@Achint08 @hemantjadon @SKrPl: Please let me know if I am missing something.

File: app/src/main/res/drawable/ic_search_24dp.xml (hunk 8 above)
Comment (SKrPl) [reply app/src/main/res/drawable/ic_search_24dp.xml:8]: It's a vector image file generated by Android Studio. In general practice, it is left as it is generated.

File: build.gradle (hunk 2 above)
Comment (hemantjadon) [reply build.gradle:8]: @abishekvashok @SKrPl Stick to Stable I guess

**Review by singhpratyush — APPROVED** [review]
I guess it we should proceed with this and make the changes when https://github.com/fossasia/loklak_scraper_js/issues/36 gets fixed.

**Comment by SKrPl** [comment]
As per the discussions and suggestions on "Automating the addition of bundled scraper in this project" (link: https://github.com/fossasia/loklak_scraper_js/issues/36), as of now bundled scraper file will be added to loklak_wok_android manually.

**Comment by singhpratyush** [comment]
@SKrPl: Also, try to make smaller PRs as they would be easy to review :sweat_smile:.

**Comment by SKrPl** [comment]
@singhpratyush I will try my best to do that from next time onwards, this was a big feature, and making different PRs could have resulted in a lot of merge conflicts, as I changed the code frequently.

**Comment by SKrPl** [comment]
@kavithaenair @abishekvashok Added new lines at the end of mentioned files and squashed the commits.

**Review by kavithaenair — APPROVED** [review]

**Review by vibhcool — APPROVED** [review]
looks good to me

**Comment by sudheesh001** [comment]
Please squash the commits.

**Comment by SKrPl** [comment]
@sudheesh001 done :smile:.

**Review by daminisatya — APPROVED** [review]
(Y)

**Pull request merged**